### DISCUSSION TITLE
What is the difference between cloning and branching?

### DISCUSSION THREAD

**Discussion by thegreendrinker** [community/community#22286] [New to GitHub]
<p>I am just learning how to use Github and SourceTree and I am really confused as to how it all works.</p>
<p>When I cloned my repo, a duplicate copy of the repo appeared in my file structure. After that, that pink branching line appeared. Is cloning essentially creating an alternate universe file history that begins at the same point as the original but then branches off in to its own thing? If so, then what is the point of “branching”? When I create a branch, it doesn’t change what appears in my file structure hierarchy at all. So what is the difference?</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/09f0edce10c68408c71d26cc5f1af52df1163ecd.png" data-download-href="/uploads/short-url/1pWu4DicawpDBB4bWHrWfia2xtP.png?dl=1" title="sourcetree_branches.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/32495526/181070062-45dc2a29-d4b1-419e-af49-48370d349b7f.png" alt="sourcetree_branches.png" data-base62-sha1="1pWu4DicawpDBB4bWHrWfia2xtP" width="690" height="454" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/09f0edce10c68408c71d26cc5f1af52df1163ecd_2_690x454.png, https://aws1.discourse-cdn.com/github/optimized/1X/09f0edce10c68408c71d26cc5f1af52df1163ecd_2_1035x681.png 1.5x, https://aws1.discourse-cdn.com/github/original/1X/09f0edce10c68408c71d26cc5f1af52df1163ecd.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/09f0edce10c68408c71d26cc5f1af52df1163ecd_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">sourcetree_branches.png</span><span class="informations">1173×773 46.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by Wabri** [comment, marked as answer]
<p>Hi <a class="mention" href="/darienspringer654">@darienspringer654</a>,</p>
<p>Welcome to Github Community Forum!</p>
<p>They are completely different. I will try to explain with an example.</p>
<p>Consider a history of a repository:</p>
<p>                 master -&gt;  A—B---C—D </p>
<p><strong>A, B, C, D are commits</strong> of this repository made on the master branch. Consider that this repository is stored in a server (like github), normaly the main repository are called <strong>origin</strong>. We want to develop some new things on this project so we <strong>clone</strong> this repository on our local pc:</p>
<p>              <strong>origin</strong> /master -&gt;  A—B---C—D </p>
<p>                                                      II   clone<br></p>
<p>              <strong>local</strong> /master -&gt;  A—B---C—D</p>
<p>After some changes and commits you will have this history:</p>
<p>              <strong>origin</strong> /master -&gt;  A—B---C—D </p>
<p>                                                      II  <br></p>
<p>              <strong>local</strong> /master -&gt;  A—B---C—D---E—F---G<br></p>
<p>And then you will push this changes to origin repository to update it:</p>
<p>              <strong>origin</strong> /master -&gt;  A—B---C—D---E—F---G</p>
<p>                                                      II   push origin<br></p>
<p>              <strong>local</strong> /master -&gt;  A—B---C—D---E—F---G</p>
<p>This works perfectly if you are the only one that are make changes in this repository and master branch, probably when you develop something you need to work with some other people and changes the same branch will cause <strong>conflicts</strong>:</p>
<p>              local_one/master -&gt;  A—B---C—D---E—F---G</p>
<p>                                                      II   push origin</p>
<p>              origin/master -&gt;  A—B---C—D</p>
<p>                                                      II   push origin<br></p>
<p>              local_two/master -&gt;  A—B---C—D---X—Y---Z</p>
<p>In this case the rule is: <strong>who comes first rule them all</strong> and the others are no longer accepted to push their changes in the master branch. Branches are the way to prevents conflicts, instead of doing all of the changes in a single branch you made that in a <strong>new different branch</strong>. Like this:</p>
<p>              <strong>local_one/new.branch.one</strong> -&gt;  A—B---C—D---E—F---G</p>
<p>                                                      II   push origin new.branch.one<br></p>
<p>                    origin/master -&gt;  A—B---C—D</p>
<p>                                                      II   push origin new.branch.two<br></p>
<p>              <strong>local_two/new.branch.two</strong> -&gt;  A—B---C—D---X—Y---Z</p>
<p>This will create a different history in the <strong>origin</strong> repository:</p>
<p>                      new.branch.one -&gt;  E—F---G</p>
<p>                                                       /<br></p>
<p>              master -&gt;  A—B---C—D</p>
<p>                                                       &lt;br&gt;</p>
<p>                     new.branch.two -&gt;  X—Y---Z</p>
<p>Now you can try to merge this branch into master with <strong>Pull Request</strong> on github or with merge command, if everything goes ok in the end you get a new history of the master branch:</p>
<p>              master -&gt;  A—B---C—D---X—Y---Z—E---F—G</p>
<p>or:</p>
<p>            master -&gt;  A—B---C—D---E—F---G—X---Y—Z</p>
<p>The order is determined by the first branch merged.<br></p>
<p>The branching is used also when a repository are made by only one developer, normally <strong>one branch correspond to a new functionality or a new update</strong> (also called patch) of the software this can help to organize the project. it also provide a better way to<strong> test new functionality</strong> of the software without touch the master code and if this functionality are no more requested you can simply delete the branch where are stored without too many efforts.</p>
<p>Branching and cloning are completly different one to another. I already explain what branches are used for, the clone is simply a copy and paste of a project. You are not going to use more than one clone for a project, but you will use a thousand of branches for a project. This is an example of a repository history with many branches:<br></p>
<p><img width="541" height="267" alt="2019-03-23_541x267.png" src="https://user-images.githubusercontent.com/12409541/181070063-8e6f731b-ab17-4a9f-9cb7-6947c5f2ee61.png"></p>
<p>The difference from copy a directory and clone a repository is simpy that the second will copy the history of the repository and create all metadata to use git protocol inside that folder.</p>
<p>I hope I have solved your doubts, <span><span>otherwise ask</span></span> for more :grin: <br></p>
<p>-Gabriele-</p>
<p>Mark helpful posts with Accept as Solution to help other users locate important info. Don’t forget to give Kudos for great contents!</p>

**Comment by thegreendrinker** [comment]
<p>Yes, you were extremely helpful. Thank you! I definitely have a better understanding of the distinction between branching and cloning now.</p>
<p>My question now is: since a branch is for testing things out that may or may not be merged in to the master, is it possible to have entire files in a branch that won’t be included in the master (unless the branch merges)? Because in Sourcetree, it doesn’t seem like I can add a branch exclusive file. I can only commit a file to the master.</p>

**Comment by Wabri** [comment]
<p>Yes, you can do whatever you want in a branch like: delete, create and transform that are files or folders. I dont know how it works sourcetree, but I think this tool it is made to manage git flow and not to edit and create new things. So you need a file manager to create a new files or directories and an editor to edit files.</p>
<p>-Gabriele-</p>
<p>Mark helpfull posts with Accept as Solution to help other users locate important info. Don’t forget to give Kudos for great contents!</p>
